### ISSUE TITLE
Nested Input Types Are Not Working (again)

### ISSUE THREAD

**Issue by SignalDancer** [graphql-java/graphql-java-tools#77]
Using version 4.1.2 and graphql-spring-boot-starter 3.9.2

schema file:

input DiscountTypeFilter{
    code_text_is: String
    id_is: Long
}

input RequestFilter{
    event_date_greater: Long
    discount_type_filter: DiscountTypeFilter
}

input MaterialFilter{
    title_contains: String
    discount_type_filter: DiscountTypeFilter
}

I get a spring boot error that is root caused here:
Caused by: com.coxautodev.graphql.tools.SchemaError: Expected type 'DiscountTypeFilter' to be a GraphQLInputType, but it wasn't!  Was a type only permitted for object types incorrectly used as an input type, or vice-versa?
	at com.coxautodev.graphql.tools.SchemaParser.determineType(SchemaParser.kt:248) ~[graphql-java-tools-4.1.2.jar:?]
	at com.coxautodev.graphql.tools.SchemaParser.determineInputType(SchemaParser.kt:236) ~[graphql-java-tools-4.1.2.jar:?]
	at com.coxautodev.graphql.tools.SchemaParser.access$determineInputType(SchemaParser.kt:48) ~[graphql-java-tools-4.1.2.jar:?]

Throwing a debugger on it the allowable types array at the time it's throwing the error are MaterialFilter and MaterialDTO which is a type object I have that instantiates fine. I'm not sure why it doesn't recognize RequestFilter or DiscountTypeFilter as additionally allowable types?

Any help is appreciated

**Comment by apottere** [comment]
Does it log a warning that those types aren't used and can be deleted?  It seems like the scanner isn't finding usages of those types so they aren't making it into the parser.

**Comment by SignalDancer** [comment]
Ah, yes, it is logging that and I missed it with all the other stuff spring boot was spitting in my console. What do I do about that to fix it? (Apologies if I missed that in the documentation)

**Comment by apottere** [comment]
It seems like a bug in the scanner that isn't finding those types properly.  Do you have a full schema you can send me?  I have a few tests in the library and I'm not sure how your schema differs from them.

**Comment by SignalDancer** [comment]
@apottere 

Sure, I've pared out the company specific items in this one, but it shows the problem fully.

type DiscountType{
    id: Long!
    name: String!
}

type Material{
    id: Long!
    request: Request!
    ISBN: String
    title: String!
    format: MaterialFormat!
    quantity: Int!
    lastModified: Long!
    lastModifiedBy: String!
}

type Request{
    id: Long!
    eventDate: Long!
    discountType: DiscountType!
}

input DiscountTypeFilter{
    name_is: String
    id_is: Long
}

input RequestFilter{
    AND: [RequestFilter!]
    OR: [RequestFilter!]
    event_date_greater: Long
    discount_type_filter: DiscountTypeFilter

}

input MaterialFilter{
    title_contains: String
    request_filter: RequestFilter
}


enum MaterialFormat{
    HARDCOPY
    EBOOK
    CERTIFICATE
    OTHER
}

type Query {
    materials(filter: MaterialFilter): [Material]!
}

schema {
    query: Query
}

**Comment by SignalDancer** [comment]
A workaround to force them to be picked up is changing the query to this:

type Query {
    materials(filter: MaterialFilter): [Material]!
    requests(filter: RequestFilter): [Request]!
    discountTypes(filter: DiscountTypeFilter): [DiscountType]!
}

Basically if they're in the root query object directly they get picked up, but if they're only nested inside another filter, they're never picked up.

**Comment by jakubkluz** [comment]
@SignalDancer
Hi, could you retest with latest SNAPSHOT version? I think that this commit  327a82bafd0316e7cc5bdccb8a6aa87ef066ea91 might fix also this issue.

**Comment by KaRkY** [comment]
Hello,

I just tried having nested input types in version 4.3.0 and it does not work.

**Comment by PMisudik** [comment]
Hello,
I have the same problem. SchemaParser does not find nested input types.

**Comment by geesen** [comment]
I'm not sure, if it is a related issue: 
I have an input type with an enum value, which produces the same failure:

    input Request {
        order: Direction
    }

    enum Direction {
        ASC
        DESC
    }

The scanner also says "Schema type was defined but can never be accessed, and can be safely deleted: Direction" although the request is defined directly in the root/query-resolver.

**Comment by mikeurbach** [comment]
I was running into this on 4.3.0, specifically using nested input objects for mutations. The workaround from @SignalDancer was enough to build my schema, but during query execution I was getting `UnrecognizedPropertyException`s from jackson. I added `@JsonProperty` annotations to the fields on my data classes for those mutations, and it is working correctly now. Hope that helps others get it running. I don't have time to put in a fix, but I can help test if someone gets around to it.

**Comment by warmwind** [comment]
@mikeurbach I have the same issue with 4.3.0. How do you fix it for input type?

**Comment by mikeurbach** [comment]
@warmwind did you try this workaround? https://github.com/graphql-java/graphql-java-tools/issues/77#issuecomment-336153637. that, and adding `@JsonProperty` annotations on the fields of the nested input class worked for me.

**Comment by PatrickSchuster** [comment]
I had a very similar problem and managed to resolve it by making sure my resolver function had the corresponding parameter.
E.g. the following schema:

```
type Query {
    findAuthors(filter: [Operand]): [Author]
}

input Operand {
    operand: String!
    filters: [AllPurposeFilter]!
}

input AllPurposeFilter {
    firstProperty: String!
    operator: String!
    secondProperty: String!
}
```
And the Java classes `Operand` and `AllPurposeFilter` containing the fields annotated with `@JsonProperty`. 
E.g. 
```
@JsonProperty("operand")
public String getOperand(){
    return this.operand;
}
```
in the `Operand` class.

The resolver method looks like that:
```
public List<Author> findAuthors(List<Operand> operands) {
   // ...
}
```

**Comment by warmwind** [comment]
@mikeurbach Thanks.  We finally do not using this lib and only use graphql-java

**Comment by CH-EricLundberg** [comment]
I am having this issue, I can perhaps work around it, but any traction of a fix?

**Comment by yeyuexia** [comment]
Hi all,
  I trace the error and find when you meet this issue, there must has something wrong in your java classes which mapping the graphql type.

**Comment by danielperovich** [comment]
I am having the same issue. When using nested input types, a warning is logged indicating that the type is not used and that it can be deleted, but afterwards an error is logged indicated that it is expected.

The workaround of exposing those input types directly in the Query type works well, but augments the exposed API. As a workaround to the latter, we can post-processs the schema (by removing the extra fields in the Query type), but still the field resolvers need to be implemented.

I am using graphql-spring-boot-starter-5.0.2, graphql-java-tools-5.2.4, graphql-java-9.2.

Is anyone else having this issue?

**Comment by oliemansm** [comment]
I've added a unit test with the test schema provided here that is still failing atm. Next step is to find the cause and solve it.

**Comment by oliemansm** [comment]
Issue is fixed. See unit test `NestedInputTypesSpec` for working example.

**Comment by zhuravskiy** [comment]
looks like still not working with nested enums and custom scalar types
please add this cases to NestedInputTypesSpec
```

**Comment by adrian-baker** [comment]
>looks like still not working with nested enums and custom scalar types

I'm seeing the same issue.

**Comment by davis91** [comment]
This is still an issue as of 5.10.0. Exposing the nested input type in a separate root query/mutation solves the issue.

**Comment by kevinsarsen** [comment]
> This is still an issue as of 5.10.0. Exposing the nested input type in a separate root query/mutation solves the issue. Though this is most likely not the most ideal workaround.

@davis91 ...could you explain "Exposing the nested input type in a separate root query/mutation" and possibly give an example? I've run into this issue and am trying to find a solution/workaround.

**Comment by joarobles** [comment]
> > This is still an issue as of 5.10.0. Exposing the nested input type in a separate root query/mutation solves the issue. Though this is most likely not the most ideal workaround.
> 
> @davis91 ...could you explain "Exposing the nested input type in a separate root query/mutation" and possibly give an example? I've run into this issue and am trying to find a solution/workaround.

He mens that the only way for the nested input type to get picked is to create a Query/Mutation that explicitly refers to it, which isn't a proper solution

**Comment by IKS-Ploetzwich** [comment]
This is still a thing.
Anyon found a proper solution, that doesn't involve implementing a dummy query/mutation?

For me, it worked with a single mutation referring to an input type. No problem.
But when I added a second mutation with a different input type it broke.
Implementing the dummy fixed it.
```
type Query {
    profile(username: String!): Profile!
    profiles(count: Int = 10, page: Int = 0): [Profile!]!
    project(id: Int!): Project!
    projects(count: Int = 10, page: Int = 0): [Project!]!
}

type Mutation {
    #Only implementing this works
    saveProfile(input: ProfileInput!): Profile!
    #When I implement this, it breaks
    saveProject(input: ProjectInput!): Project!
}

#### Types ####
type Profile {
    id: ID!
    name: String!
    img: String!
    email: String!
    username: String!
    skills: [Skill!]
    projects: [Project!]
}

type Project {
    id: Int!
    name: String!
    description: String
    technologies: [Technology!]
    profiles: [Profile!]
}

type Skill {
    id: Int!
    name: String!
    level: Int!
}

type Technology {
    id: Int!
    name: String!
}


#### Inputs ####
input SkillInput {
    id: Int
    name: String!
    level: Int!
}

input TechnologyInput {
    id: Int
    name: String!
}

input ProfileInput {
    id: ID!
    firstName: String!
    lastName: String!
    email: String!
    username: String!
    img: String!
    skills: [SkillInput!]
    projects: [Int!]
}

input ProjectInput {
    id: Int!
    name: String
    description: String
    technologies: [TechnologyInput!]
    profiles: [ID!]
}
```

**Comment by IKS-Ploetzwich** [comment]
So, I think I have a lead here.

**TL;dr check if all field names in your POJO match those in your schema definition**

After endlessly reviewing my code I found that I made a typo in the corresponding java class.
The relevant SDL code would be:
```
input ProjectInput {
    id: Int
    name: String
    description: String
    technologies: [TechnologyInput!]
    profiles: [ID!]
}

input TechnologyInput {
    id: Int
    name: String!
}
```
which is corresponding to my model class for ProjectInput
```
@Data
@AllArgsConstructor
@NoArgsConstructor
public class ProjectInput {
    private int id;
    private String name;
    private String description;
    // The typo was in this line. I miss typed "technolgies"
    private List<TechnologyInput> technologies;
    private List<String> profiles;
}
```
As you can see, with the typo the name of the field it doesn't match the property in my SDL definition anymore.
After fixing the typo my code compiles without error.

My hypothesis would be, that the variables need to be the same name as stated in the schema definition.
Is this intended behavior? Or did I simply miss something?

I hope this might help some people.

**Comment by vojtapol** [comment]
That sounds like a reasonable expectation. I have opened this to track the problem:  https://github.com/graphql-java-kickstart/graphql-java-tools/issues/new

**Comment by ybhatnagar** [comment]
@IKS-Ploetzwich  This was the exact issue for me as well!!

**Comment by sbmaggarwal** [comment]
> looks like still not working with nested enums and custom scalar types

I'm seeing the same issue.

**Comment by vojtapol** [comment]
> > looks like still not working with nested enums and custom scalar types
> 
> I'm seeing the same issue.

Can you give us more information how to reproduce this? Ideally a PR with a failing test case.

**Comment by mryan43** [comment]
@IKS-Ploetzwich you are my hero ! thanks a lot I had a typo as well.

The error message in this case is very misleading. Could it be improved to indicate that the field specified in the schema cannot be found in the corresponding java class instead of saying  "Expected type 'MyFilter' to be a GraphQLInputType, but it wasn't!" ?
